Vimeo download

jezhiggins opened the pull request:
If components.sjon references a Vimeo hosted video, download it and
package into the OBB

davetaz commented:
Hi Jez,
I guess we were both working over Easter. 
I have just merged a large amount of changes I did into the master here which includes the video downloading. I had a look at your pull request but it appeared to be synchronous? In my version I used execSync with a new script to download the videos. Also I have updated the code as we also need to build a 'fat' ios version. 

I think this is almost there. There is one more major change needed to the package identifiers as follows: 

IOS built as is with vimeo in the cloud: <package.id>-lite
IOS built with videos built in (fat): <package.id>

Android built with obb file: <package.id>
Android built with as is with vimeo in the cloud: <package.id>-lite
Android build with videos built in (fat): <package.id>-full

I think this involves a change to the cordova config up font which I'll look at today probably. It would also be nice if the IOS build process output 2 directories (1 for each) which can be loaded into xCode. 

I have yet to test the signing in Android, which I'm going to look at now with this version (on master)

jezhiggins commented:
Hi Dave,

> I guess we were both working over Easter.
> I have just merged a large amount of changes I did into the master here
> which includes the video downloading. I had a look at your pull request
> but it appeared to be synchronous? 

Yes, and no :)

The vimeo downloader library reports what it's downloaded using
callbacks, and that didn't fit with the strictly linear mode I had in
the script at that point.  So, I've reworked it interally to use
promises to wrap up the callbacks, and reimpose a sequence.

At the Cordova entrypoint, it again wraps things in a promise (this time
from the Q library) which is returned to Corddova.  Cordova waits for
that promise to resolve (which happens once everything has downloaded).
So, internally it's built with async callbacks and promises, while from
the outside it appears as a sync step in the Cordova build pipeline.

> In my version I used execSync with a
> new script to download the videos. Also I have updated the code as we
> also need to build a 'fat' ios version.
> 
> I think this is almost there. There is one more major change needed to
> the package identifiers as follows:
> 
> IOS built as is with vimeo in the cloud: <package.id>-lite
> IOS built with videos built in (fat): <package.id>
> 
> Android built with obb file: <package.id>
> Android built with as is with vimeo in the cloud: <package.id>-lite
> Android build with videos built in (fat): <package.id>-full

This should be straightforward enough, I think, if we split the
package_video.js into two, one to download videos and add them to the
source, and one to extract and package the obb

The order of things would then be
   build cloud
   add download videos hook
   build fat
   remove download videos, add extract and package video
   build lite

If I get a chance today, I'll see if I can sort that out.

> I think this involves a change to the cordova config up font which I'll
> look at today probably. It would also be nice if the IOS build process
> output 2 directories (1 for each) which can be loaded into xCode.

There's a step in cordova_build which looks for output files and moves
them up into our root directory.  At the moment it's only used to grab
the Android .apk and .obb files.  You could repurpose this to move the
IOS output in a similar way.

> I have yet to test the signing in Android, which I'm going to look at
> now with this version (on master)

As I said in my previous email, we might need to pull out another
password parameter.

Jez
